From 382533ac7169445e1be0344a2c0a74733584777e Mon Sep 17 00:00:00 2001
From: Adrian Prantl <adrian@llnl.gov>
Date: Mon, 9 Jul 2012 10:09:08 -0700
Subject: [PATCH 03/11] add support for configurable field name during assignment

---
 compiler/AST/expr.cpp |    7 +++++++
 1 files changed, 7 insertions(+), 0 deletions(-)

diff --git a/compiler/AST/expr.cpp b/compiler/AST/expr.cpp
index db36fe6..fea3df6 100644
--- a/compiler/AST/expr.cpp
+++ b/compiler/AST/expr.cpp
@@ -2744,6 +2744,13 @@ GenRet CallExpr::codegen() {
       codegenAssign(dst, alloced);
       break;
     }
+    case PRIM_REF_BORROW:
+    {
+      codegenCall("_REF_BORROW", get(1),
+		  getDataClassType(get(1)->typeInfo()->symbol),
+		  get(2), get(3), get(4));
+      break;  
+    }
     case PRIM_GPU_ALLOC:
     {
       if (get(1)->typeInfo()->symbol->hasFlag(FLAG_WIDE_CLASS)) {
-- 
1.7.0.4

