From cffc838537b99afedae301d2f501b708614ecc31 Mon Sep 17 00:00:00 2001
From: Adrian Prantl <adrian@llnl.gov>
Date: Wed, 25 Jul 2012 11:37:04 -0700
Subject: [PATCH 05/11] add library init function

---
 runtime/src/main.c |    8 +++-----
 1 files changed, 3 insertions(+), 5 deletions(-)

diff --git a/runtime/src/main.c b/runtime/src/main.c
index 9452677..dc0f57c 100644
--- a/runtime/src/main.c
+++ b/runtime/src/main.c
@@ -5,10 +5,8 @@
 #include "chplcast.h"
 #include "chplcgfns.h"
 #include "chpl-comm.h"
+#include "chplinit.h"
 #include "chplexit.h"
-#include "chplio.h"
-#include "chpl-mem.h"
-#include "chplmemtrack.h"
 #include "chpl-tasks.h"
 #include "config.h"
 #include "error.h"
@@ -165,11 +163,11 @@ int main(int argc, char* argv[]) {
   // The call to chpl_comm_barrier makes sure that all locales are listening
   // before an attempt is made to run tasks "on" them.
 
-  if (chpl_localeID == 0) {      // have locale #0 run the user's main function
+  if (chpl_localeID == 0) { // have locale #0 run the user's main function
 
     chpl_task_callMain(chpl_main);
   }
 
   chpl_exit_all(0);         // have everyone exit
-  return 0;                  // should never get here
+  return 0;                 // should never get here
 }
-- 
1.7.0.4

