From 873ac147a8d84302bfa42bbc6a5dd0e96c1e243e Mon Sep 17 00:00:00 2001
From: Adrian Prantl <adrian@llnl.gov>
Date: Wed, 7 Nov 2012 11:03:42 -0800
Subject: [PATCH 09/11] add endCount argument to _waitEndCount call

---
 compiler/passes/buildDefaultFunctions.cpp |    4 ++--
 1 files changed, 2 insertions(+), 2 deletions(-)

diff --git a/compiler/passes/buildDefaultFunctions.cpp b/compiler/passes/buildDefaultFunctions.cpp
index fd4e065..82ead55 100644
--- a/compiler/passes/buildDefaultFunctions.cpp
+++ b/compiler/passes/buildDefaultFunctions.cpp
@@ -359,13 +359,13 @@ static void build_chpl_entry_points(void) {
     // We have to initialize the main module explicitly.
     // It will initialize all the modules it uses, recursively.
     chpl_gen_main->insertAtHead(new CallExpr(mainModule->initFn));
-
+     
     VarSymbol* endCount = newTemp("_endCount");
     chpl_gen_main->insertAtHead(new CallExpr("chpl_startTrackingMemory"));
     chpl_gen_main->insertAtHead(new CallExpr(PRIM_SET_END_COUNT, endCount));
     chpl_gen_main->insertAtHead(new CallExpr(PRIM_MOVE, endCount, new CallExpr("_endCountAlloc")));
     chpl_gen_main->insertAtHead(new DefExpr(endCount));
-    chpl_gen_main->insertBeforeReturn(new CallExpr("_waitEndCount"));
+    chpl_gen_main->insertBeforeReturn(new CallExpr("_waitEndCount", endCount));
     //chpl_gen_main->insertBeforeReturn(new CallExpr("_endCountFree", endCount));
   }
   chpl_gen_main->insertAtHead(new CallExpr("chpl_gen_init"));
-- 
1.7.0.4

