commit e1e5f348671bab4060acad3a7ef3b11c60c33de3
Author: Adrian Prantl <adrian@llnl.gov>
Date:   Fri Nov 30 15:57:31 2012 -0800

    synch with upstream

diff --git a/compiler/passes/codegen.cpp b/compiler/passes/codegen.cpp
index 579f014..5bde8cc 100644
--- a/compiler/passes/codegen.cpp
+++ b/compiler/passes/codegen.cpp
@@ -107,14 +107,7 @@ static void legalizeName(Symbol* sym) {
   }
 
   // Add chpl_ to operator names.
-  // hilde sez:  This is very kludgy.  What do we really mean?
-  if ((!strncmp("chpl_", sym->cname, 5) && 
-       (strcmp("chpl_gen_main", sym->cname) &&
-	strcmp("chpl_gen_init", sym->cname) &&
-	strcmp("chpl_gen_shutdown", sym->cname) &&
-	strcmp("chpl_user_main", sym->cname)) 
-       && sym->cname[5] != '_') ||
-      (sym->cname[0] == '_' && (sym->cname[1] == '_' || (sym->cname[1] >= 'A' && sym->cname[1] <= 'Z')))) {
+  if (sym->cname[0] == '_' && (sym->cname[1] == '_' || (sym->cname[1] >= 'A' && sym->cname[1] <= 'Z'))) {
     sym->cname = astr("chpl__", sym->cname);
   }
 }
diff --git a/compiler/resolution/callDestructors.cpp b/compiler/resolution/callDestructors.cpp
index a22a713..b683f05 100644
--- a/compiler/resolution/callDestructors.cpp
+++ b/compiler/resolution/callDestructors.cpp
@@ -341,7 +341,6 @@ static void insertGlobalAutoDestroyCalls() {
   if( ! fNoInternalModules ) {
     chpl_gen_main->defPoint->insertBefore(new DefExpr(fn));
     chpl_gen_main->insertBeforeReturnAfterLabel(new CallExpr(fn));
-    chpl_gen_shutdown->defPoint->insertBefore(new DefExpr(fn));
     chpl_gen_shutdown->insertBeforeReturnAfterLabel(new CallExpr(fn));
   }
   forv_Vec(DefExpr, def, gDefExprs) {
diff --git a/runtime/src/chplinit.c b/runtime/src/chplinit.c
index 1b8d0ce..af74c45 100644
--- a/runtime/src/chplinit.c
+++ b/runtime/src/chplinit.c
@@ -39,7 +39,7 @@ void printAdditionalHelp(void) {
 }
 
 
-static void recordExecutionCommand(int argc, char *argv[]) {
+void recordExecutionCommand(int argc, char *argv[]) {
   int i, length = 0;
   for (i = 0; i < argc; i++) {
     length += strlen(argv[i]) + 1;
@@ -230,4 +230,4 @@ int chpl_init_library(int argc, char* argv[]) {
   }
 
   return chpl_localeID;
-}
\ No newline at end of file
+}
